### ISSUE TITLE
marginal_counts erases information in input result object

### ISSUE THREAD

**Issue by aeddins-ibm** [Qiskit/qiskit-terra#4912]
### Information

- **Qiskit Terra version**: 0.14.2
- **Python version**: 3.7
- **Operating system**: MacOS

### What is the current behavior?
The marginal_counts function in result/utils.py modifies the input result without first creating a copy, so that the original counts data in the job object is destroyed. See image.

![image](https://user-images.githubusercontent.com/60495383/89957579-9f52f400-dbec-11ea-92f4-fc5ad77d3a35.png)

### Steps to reproduce the problem
```
from qiskit import QuantumCircuit, schedule, transpile
from qiskit import IBMQ
from qiskit import assemble
from qiskit.result import utils
from copy import deepcopy

IBMQ.load_account()
# provider = IBMQ.get_provider(hub='your-hub', group='your-group', project='your-project')


from qiskit import Aer
backend=provider.get_backend('your-backend')
qc = QuantumCircuit(2,2)
qc.x(0)
qc.measure([0, 1], [0, 1])
qc = transpile(qc, backend=provider.get_backend('your-backend'), initial_layout=[4,2]) # initial_layout is arbitrary here
print(qc.draw())
sched = schedule(qc, backend=provider.get_backend('your-backend'))
qobj = assemble(sched, backend=provider.get_backend('your-backend'), shots=1024)
job = backend.run(qobj)

print('running...')
results_all = job.result()
print('\ninitially, job object has count info for all qubits:')
print(results_all.get_counts())

print('\nresult for only Q9 after marginalizing:')
result = utils.marginal_counts(results_all,indices=[0])
# result = utils.marginal_counts(deepcopy(results_all),indices=[0])
# commented code above fixes the issue, so the deepcopy could be moved inside marginal_counts; if memory is a concern an "in_place" kwarg or equivalent could also be added to disable the deep_copy
print(result.get_counts())

print('\ncounts information in original job is also marginalized, so we have inadvertently erased counts information on the other qubits from memory:')
results_all = job.result()
results_all.get_counts()
```

### What is the expected behavior?
Marginal counts should act on an independent copy of the input result/counts object, returning the marginalized counts without modifying the input.

### Suggested solutions
1) Add this at the top of utils.py:
`import copy`
Add this at the start of marginal_counts:
`result = copy.deepcopy(result)`
This breaks the connection to the original input result object so we don't accidentally modify it.
Need to verify this works for both dict and Result type inputs for the result argument.

2) For users concerned about memory, could add a kwarg "in_place=False", so the new code in marginal_counts reads:
if not in_place:
    result = copy.deepcopy(result)

3) While we're at it, update the docstring description of the indices argument. Currently it says: "The bit positions of interest to marginalize over." This description further confuses an already confusing process, since standard terminology is that the bit positions of interest are the ones NOT marginalized over. (e.g. Wikipedia page on Marginal Distribution: _The distribution of the marginal variables ... is obtained by marginalizing ... over the distribution of the variables being discarded ...._). Update to read "The bit positions of interest to keep (all others will be marginalized out)", after confirming that is the correct interpretation.

**Comment by abhik-99** [comment]
Hey, How do I contribute to this?

**Comment by coder-mtv** [comment]
Can I work on this?

**Comment by taalexander** [comment]
Please let me know if you wish me to assign you this issue @abhik-99. If not I can assign to @coder-mtv.

**Comment by abhik-99** [comment]
Sure @taalexander ! I can do this.

**Comment by kdk** [comment]
Closing as resolved by #4929 .
